File: scripts/release.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.11"
# dependencies = []
# ///
"""The release tag algebra: two tag shapes, one order, no ambiguity.

Every binary this project ships is named by a git tag, and there are exactly two
kinds:

    v0.4.2                                       a release, cut by hand
    v0.4.3-dev.2026-08-24.11-30-00.a1b2c3d4e5f6  a pre-release, cut by main

**Both are valid SemVer, and that is the whole point.** SemVer 2.0.0 §11 makes a
pre-release sort *before* the release it shares a core version with, so
`v0.4.3-dev...` precedes `v0.4.3` — which is why a pre-release is numbered for
the patch that has not happened yet rather than the one that has. A consumer
sorting tags gets "newest" for free, and gets it right.

Two properties of the stamp carry that ordering and both are load-bearing:

* **ISO date, zero-padded time.** SemVer compares dot-separated identifiers
  left to right, and any identifier containing a hyphen is compared as ASCII.
  `2026-08-24` and `11-30-00` therefore sort chronologically *because* they are
  zero-padded and big-endian. `2026-8-24` or `9-30-00` would sort wrongly and
  nothing would notice until two builds an hour apart came back in the wrong
  order.
* **The commit last.** It is the tiebreaker for two builds in the same second,
  and it is what makes a tag say which tree it came from without a lookup.

One caveat, recorded rather than fixed: SemVer ranks an all-digit identifier
below any alphanumeric one, so a commit abbreviating to twelve digits would sort
below its siblings. That can only reorder two builds stamped in the same second,
which is not a distinction anything here draws.

    scripts/release.py --next          the next pre-release tag, from git
    scripts/release.py --check v0.4.2  gate a release tag against Cargo.toml
    scripts/release.py --selftest      this module's doctests, which are the laws

**Run it through `uv`.** The PEP 723 block above is the whole environment.
"""

from __future__ import annotations

import argparse
import contextlib
import doctest
import enum
import io
import re
import subprocess
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import NamedTuple

import tomllib

ROOT = Path(__file__).resolve().parent.parent

#: How much of the commit goes in a tag. Twelve is unambiguous for any tree this
#: project will have and still reads in one glance.
COMMIT_CHARS = 12


class Exit(enum.IntEnum):
    """Exit statuses, as a closed sum rather than scattered integers."""

    OK = 0
    ERROR = 1
    MALFORMED = 2


class Version(NamedTuple):
    """A SemVer core version.

    A `NamedTuple` because tuple ordering *is* SemVer's core ordering — compare
    major, then minor, then patch, numerically — so the order comes from the
    representation rather than from a method that could disagree with it.

    >>> Version(0, 4, 2) < Version(0, 4, 10) < Version(0, 5, 0) < Version(1, 0, 0)
    True
    """

    major: int
    minor: int
    patch: int

    def __str__(self) -> str:
        return f"{self.major}.{self.minor}.{self.patch}"

    def bump_patch(self) -> Version:
        """The next patch version.

        >>> str(Version(0, 4, 2).bump_patch())
        '0.4.3'
        """
        return Version(self.major, self.minor, self.patch + 1)


class Release(NamedTuple):
    """A version someone decided to ship. Cut by hand, never by CI.

    >>> str(Release(Version(1, 2, 3)))
    'v1.2.3'
    """

    version: Version

    def __str__(self) -> str:
        return f"v{self.version}"


class PreRelease(NamedTuple):
    """A build of `main`, named for the patch it is heading toward.

    `commit` is the abbreviated hash of the tree it was built from; `stamp` is
    when. Both are rendered into the tag, so the tag alone answers "which tree,
    and when" without a lookup.

    >>> at = datetime(2026, 8, 24, 11, 30, 0, tzinfo=timezone.utc)
    >>> str(PreRelease(Version(0, 4, 3), at, "a1b2c3d4e5f6"))
    'v0.4.3-dev.2026-08-24.11-30-00.a1b2c3d4e5f6'
    """

    version: Version
    stamp: datetime
    commit: str

    def __str__(self) -> str:
        day = self.stamp.strftime("%Y-%m-%d")
        # Hyphens, not colons: SemVer's identifier alphabet is `[0-9A-Za-z-]`,
        # and a colon would make the whole tag invalid rather than merely ugly.
        time = self.stamp.strftime("%H-%M-%S")
        return f"v{self.version}-dev.{day}.{time}.{self.commit}"


#: A release tag and nothing else. Anchored at both ends, so `v1.2.3-dev.…`
#: does not match: the two shapes are disjoint by construction rather than by
#: the order the parser tries them in.
RELEASE = re.compile(r"^v(\d+)\.(\d+)\.(\d+)$")

#: A pre-release tag. The stamp's field widths are fixed, which is what the
#: ordering law depends on — see the module docstring.
PRERELEASE = re.compile(
    r"^v(\d+)\.(\d+)\.(\d+)-dev\.(\d{4}-\d{2}-\d{2})\.(\d{2}-\d{2}-\d{2})\.([0-9a-f]+)$"
)


def parse(tag: str) -> Release | PreRelease | None:
    """The one way a tag string becomes a tag.

    `None` for anything that is not one of the two shapes, so a tag someone
    typed by hand is refused here rather than reaching a build that names an
    artefact after it.

    >>> parse("v1.2.3")
    Release(version=Version(major=1, minor=2, patch=3))
    >>> parse("v0.4.3-dev.2026-08-24.11-30-00.a1b2c3d4e5f6").commit
    'a1b2c3d4e5f6'

    Round-tripping is the law that keeps the parser and the renderer honest.

    >>> at = datetime(2026, 8, 24, 11, 30, 0, tzinfo=timezone.utc)
    >>> both = [Release(Version(1, 2, 3)), PreRelease(Version(0, 4, 3), at, "abc123abc123")]
    >>> all(parse(str(tag)) == tag for tag in both)
    True

    Everything else is refused, including the shapes that look close.

    >>> [parse(bad) for bad in ("1.2.3", "v1.2", "v1.2.3-rc1", "v1.2.3-dev.2026-8-24.1-2-3.ab")]
    [None, None, None, None]
    """
    if (release := RELEASE.match(tag)) is not None:
        return Release(Version(*map(int, release.groups())))
    if (pre := PRERELEASE.match(tag)) is not None:
        major, minor, patch, day, time, commit = pre.groups()
        stamp = datetime.strptime(f"{day} {time}", "%Y-%m-%d %H-%M-%S").replace(
            tzinfo=timezone.utc
        )
        return PreRelease(Version(int(major), int(minor), int(patch)), stamp, commit)
    return None


def precedence(tag: Release | PreRelease) -> tuple:
    """SemVer 2.0.0 §11 precedence, as a sort key.

    The `1` and `0` are the rule that a release outranks any pre-release of the
    same core version. Everything after them only ever compares two
    pre-releases, where the stamp's fixed widths make ASCII order chronological.

    A pre-release sorts below the release it is heading toward:

    >>> at = datetime(2026, 8, 24, 11, 30, 0, tzinfo=timezone.utc)
    >>> precedence(PreRelease(Version(0, 4, 3), at, "aaa")) < precedence(Release(Version(0, 4, 3)))
    True

    And above the release it was cut from, which is what `bump_patch` buys:

    >>> precedence(Release(Version(0, 4, 2))) < precedence(PreRelease(Version(0, 4, 3), at, "aaa"))
    True

    Later builds sort later — the property the zero padding exists for. An
    unpadded `9-30-00` would sort *above* `11-30-00` and break this:

    >>> morning = datetime(2026, 8, 24, 9, 30, 0, tzinfo=timezone.utc)
    >>> noon = datetime(2026, 8, 24, 11, 30, 0, tzinfo=timezone.utc)
    >>> precedence(PreRelease(Version(1, 0, 0), morning, "aaa")) < \
            precedence(PreRelease(Version(1, 0, 0), noon, "aaa"))
    True
    """
    match tag:
        case Release(version):
            return (*version, 1)
        case PreRelease(version, stamp, commit):
            # The two stamp identifiers joined, which orders identically to
            # comparing them one at a time *because* both are fixed width.
            return (*version, 0, stamp.strftime("%Y-%m-%d.%H-%M-%S"), commit)
        case _:  # pragma: no cover - the sum is closed
            raise TypeError(f"not a tag: {tag!r}")


def base(releases: list[Version], crate: Version) -> Version:
    """The core version the next pre-release carries.

    **One `max` over a total order, rather than a branch.** The next pre-release
    heads toward the patch after the newest release — but `Cargo.toml` is also a
    declaration of where the version is going, and before the first release it
    is the only one there is. Taking the larger of the two is correct in every
    case without asking which source is authoritative.

    Before any release, the crate's own version is the answer:

    >>> str(base([], Version(0, 1, 0)))
    '0.1.0'

    After one, the patch above it — so the pre-release sorts above the release
    it followed and below the release it precedes:

    >>> str(base([Version(0, 1, 0)], Version(0, 1, 0)))
    '0.1.1'

    A crate version raised ahead of the tags wins, which is how a minor bump
    reaches pre-releases before its release exists:

    >>> str(base([Version(0, 1, 0)], Version(0, 2, 0)))
    '0.2.0'

    And tags ahead of a lagging `Cargo.toml` also win, so a forgotten bump
    cannot make two builds claim one version:

    >>> str(base([Version(0, 1, 0), Version(0, 3, 0)], Version(0, 1, 0)))
    '0.3.1'
    """
    return max([crate, *(version.bump_patch() for version in releases)])


def crate_version(manifest: Path) -> Version:
    """The version `Cargo.toml` declares.

    One number for the C ABI and the Rust crate both, as `api/stability.md`
    promises: they ship from one repository and one build.
    """
    declared = tomllib.loads(manifest.read_text())["package"]["version"]
    match parse(f"v{declared}"):
        case Release(version):
            return version
        case _:
            raise ValueError(f"Cargo.toml declares no release version: {declared!r}")


def git(*arguments: str) -> str:
    return subprocess.run(
        ["git", *arguments], cwd=ROOT, check=True, capture_output=True, text=True
    ).stdout.strip()


def released() -> list[Version]:
    """Every release tag in the repository, ignoring everything else.

    Pre-releases are deliberately not counted: they are cut by every push, and
    letting them raise the base would make the version climb with commit volume
    rather than with intent.
    """
    return [
        tag.version
        for line in git("tag", "--list", "v*").splitlines()
        if isinstance(tag := parse(line.strip()), Release)
    ]


def smoke() -> int:
    """Run `main` against this repository, for the same reason `android.py`
    does: doctests prove the algebra and nothing about the argument parser or
    the two files it reads.

    The gate is checked from both sides — the crate's own version is accepted,
    a version that is not it is refused, and a string that is not a release tag
    is refused before either is consulted.
    """
    declared = crate_version(ROOT / "Cargo.toml")
    wrong = declared.bump_patch()
    failures = 0

    def gate(tag: str) -> int:
        """`--check`, with its diagnostics swallowed: three of the five cases
        below are *expected* refusals, and a passing selftest should be quiet."""
        with (
            contextlib.redirect_stdout(io.StringIO()),
            contextlib.redirect_stderr(io.StringIO()),
        ):
            return main(["--check", tag])

    for complaint, ok in [
        (
            "--next did not produce a pre-release tag",
            isinstance(parse(next_tag()), PreRelease),
        ),
        (f"--check rejected v{declared}", gate(f"v{declared}") == Exit.OK),
        (
            f"--check accepted v{wrong}",
            gate(f"v{wrong}") == Exit.MALFORMED,
        ),
        (
            "--check accepted a pre-release tag",
            gate(next_tag()) == Exit.MALFORMED,
        ),
        (
            "--check accepted a bare version",
            gate(str(declared)) == Exit.MALFORMED,
        ),
    ]:
        if not ok:
            print(f"release: {complaint}", file=sys.stderr)
            failures += 1

    return failures


def next_tag() -> str:
    """The pre-release tag this commit would be published under."""
    return str(
        PreRelease(
            base(released(), crate_version(ROOT / "Cargo.toml")),
            datetime.now(timezone.utc),
            git("rev-parse", f"--short={COMMIT_CHARS}", "HEAD"),
        )
    )


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    what = parser.add_mutually_exclusive_group(required=True)
    what.add_argument("--next", action="store_true", help="the next pre-release tag")
    what.add_argument("--check", metavar="TAG", help="gate a release tag")
    what.add_argument(
        "--selftest", action="store_true", help="run the doctests and exit"
    )
    arguments = parser.parse_args(argv)

    if arguments.selftest:
        results = doctest.testmod(verbose=False)
        failures = smoke()
        print(
            f"release: {results.attempted} doctests, {results.failed} failed; "
            f"the gate smoke-tested, {failures} failed",
            file=sys.stderr,
        )
        return Exit.ERROR if results.failed or failures else Exit.OK

    manifest = ROOT / "Cargo.toml"

    if arguments.next:
        print(next_tag())
        return Exit.OK

    # The gate. A release is the one tag a human types, so it is the one that
    # can disagree with the tree it names — and a released artefact whose
    # version does not match the crate it was built from is discovered by a
    # downstream consumer, months later, as a mystery.
    tag = parse(arguments.check)
    if not isinstance(tag, Release):
        print(
            f"{arguments.check!r} is not a release tag; releases are vMAJOR.MINOR.PATCH",
            file=sys.stderr,
        )
        return Exit.MALFORMED

    declared = crate_version(manifest)
    if tag.version != declared:
        print(
            f"tag {tag} disagrees with Cargo.toml {declared}: bump one to match the other",
            file=sys.stderr,
        )
        return Exit.MALFORMED

    print(tag.version)
    return Exit.OK


if __name__ == "__main__":
    sys.exit(main())

File: xtask/src/release.rs
//! The release tag algebra.
//!
//! Published binaries use one of two tag shapes:
//!
//! ```text
//! v0.4.2                                        a release, cut by hand
//! v0.4.3-dev.2026-08-24.11-30-00.g1a2b3c4        a pre-release, cut by main
//! ```
//!
//! Both are valid SemVer. SemVer 2.0.0 §11 sorts a pre-release below the
//! release with the same core version, so the development tag falls between
//! `v0.4.2` and `v0.4.3` when tags are sorted.
//!
//! Three representation choices preserve that ordering:
//!
//! * [`Version`] uses field order for SemVer precedence, so derived `Ord` is
//!   sufficient.
//! * [`Stamp`] is fixed-width and zero-padded, making lexical order match
//!   chronological order.
//! * [`Sha`] has a leading `g`, making every hash identifier alphanumeric under
//!   SemVer's comparison rules.

use std::fmt;

use time::OffsetDateTime;

/// A release triple.
///
/// Field order matches SemVer precedence, so derived `Ord` is sufficient.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub struct Version {
    major: u64,
    minor: u64,
    patch: u64,
}

/// The identity for a repository with no releases.
pub const ORIGIN: Version = Version {
    major: 0,
    minor: 0,
    patch: 0,
};

impl Version {
    /// The next patch version.
    #[must_use]
    pub const fn successor(self) -> Self {
        Self {
            patch: self.patch + 1,
            ..self
        }
    }

    /// Parses a strict `vMAJOR.MINOR.PATCH` release tag.
    ///
    /// Pre-release tags return `None` and do not participate in [`resolve`].
    #[must_use]
    pub fn parse_tag(tag: &str) -> Option<Self> {
        Self::parse_triple(tag.strip_prefix('v')?)
    }

    /// Parses three SemVer numeric identifiers.
    ///
    /// Leading zeroes are rejected because SemVer forbids them.
    #[must_use]
    pub fn parse_triple(text: &str) -> Option<Self> {
        let mut fields = text.split('.').map(numeric_identifier);
        let (major, minor, patch) = (fields.next()??, fields.next()??, fields.next()??);
        fields.next().is_none().then_some(Self {
            major,
            minor,
            patch,
        })
    }
}

impl fmt::Display for Version {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}.{}.{}", self.major, self.minor, self.patch)
    }
}

/// `u64::from_str` accepts `+` and leading zeroes; SemVer accepts neither.
fn numeric_identifier(text: &str) -> Option<u64> {
    let well_formed = !text.is_empty()
        && text.bytes().all(|b| b.is_ascii_digit())
        && !(text.len() > 1 && text.starts_with('0'));
    well_formed.then(|| text.parse().ok())?
}

/// A UTC instant rendered `yyyy-mm-dd.hh-mm-ss`.
///
/// Fixed width and zero padding make lexical order match chronological order.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub struct Stamp(OffsetDateTime);

impl Stamp {
    /// Constructs a test timestamp from Unix seconds.
    #[cfg(test)]
    #[must_use]
    pub fn from_unix(seconds: i64) -> Option<Self> {
        OffsetDateTime::from_unix_timestamp(seconds).ok().map(Self)
    }

    #[must_use]
    pub fn now() -> Self {
        Self(OffsetDateTime::now_utc())
    }
}

impl fmt::Display for Stamp {
    /// Field-by-field formatting keeps the ordering-dependent zero padding
    /// explicit.
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let (date, time) = (self.0.date(), self.0.time());
        write!(
            f,
            "{:04}-{:02}-{:02}.{:02}-{:02}-{:02}",
            date.year(),
            u8::from(date.month()),
            date.day(),
            time.hour(),
            time.minute(),
            time.second(),
        )
    }
}

/// Seven hex digits of the commit, rendered with a `g` prefix.
///
/// SemVer compares all-digit identifiers numerically and ranks them below
/// alphanumeric identifiers; the prefix avoids that ordering for hashes such
/// as `0012345`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Sha([u8; 7]);

impl Sha {
    /// Parses a full or abbreviated hexadecimal object name, keeping seven
    /// digits.
    #[must_use]
    pub fn parse(full: &str) -> Option<Self> {
        let bytes = full.as_bytes();
        (bytes.len() >= 7 && bytes.iter().all(u8::is_ascii_hexdigit))
            .then(|| Self(bytes[..7].try_into().expect("length checked above")))
    }
}

impl fmt::Display for Sha {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "g{}",
            str::from_utf8(&self.0).expect("hex digits are UTF-8")
        )
    }
}

/// What triggered a publish.
///
/// The variants make a release event without a version unrepresentable.
/// `Release` carries a parsed [`Version`], so downstream code cannot receive an
/// unvalidated tag string.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Event {
    Push,
    Release(Version),
}

/// What is being published.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Publish {
    Release(Version),
    Pre {
        version: Version,
        stamp: Stamp,
        sha: Sha,
    },
}

impl Publish {
    /// The git tag and published artefact name.
    #[must_use]
    pub fn tag(&self) -> String {
        match self {
            Self::Release(version) => format!("v{version}"),
            Self::Pre {
                version,
                stamp,
                sha,
            } => format!("v{version}-dev.{stamp}.{sha}"),
        }
    }

    /// The core version used in archive names.
    #[must_use]
    pub const fn version(&self) -> Version {
        match self {
            Self::Release(version) | Self::Pre { version, .. } => *version,
        }
    }

    /// Whether GitHub should mark this a pre-release and exclude it from
    /// `Latest`.
    #[must_use]
    pub const fn is_prerelease(&self) -> bool {
        matches!(self, Self::Pre { .. })
    }
}

/// Resolves an event into a publication.
///
/// Releases use the version named by their tag. Pushes use the patch after the
/// newest release, or [`ORIGIN`] when no release exists. No manifest version is
/// consulted, so the tag is the sole release-version source.
///
/// O(n) in the released-tag count.
pub fn resolve(
    event: Event,
    released: impl IntoIterator<Item = Version>,
    now: Stamp,
    sha: Sha,
) -> Publish {
    match event {
        Event::Release(version) => Publish::Release(version),
        Event::Push => Publish::Pre {
            version: released.into_iter().max().unwrap_or(ORIGIN).successor(),
            stamp: now,
            sha,
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// 2026-08-25 11:30:00 UTC.
    const NOON: i64 = 1_787_657_400;

    fn version(text: &str) -> Version {
        Version::parse_triple(text).expect("a triple")
    }

    fn sha() -> Sha {
        Sha::parse("1a2b3c4d5e6f7890abcdef1234567890abcdef12").expect("hex")
    }

    fn stamp(seconds: i64) -> Stamp {
        Stamp::from_unix(seconds).expect("in range")
    }

    fn pre(released: &[&str], at: i64) -> Publish {
        resolve(
            Event::Push,
            released.iter().filter_map(|t| Version::parse_tag(t)),
            stamp(at),
            sha(),
        )
    }

    /// Published tag shape, including its timestamp and commit.
    #[test]
    fn a_pre_release_names_its_time_and_its_commit() {
        assert_eq!(
            pre(&["v0.4.2"], NOON).tag(),
            "v0.4.3-dev.2026-08-25.11-30-00.g1a2b3c4"
        );
    }

    /// SemVer ordering places a pre-release between its adjacent releases.
    #[test]
    fn a_pre_release_sorts_between_the_releases_it_lies_between() {
        let before = Publish::Release(version("0.4.2"));
        let middle = pre(&["v0.4.2"], NOON);
        let after = Publish::Release(version("0.4.3"));

        assert_eq!(middle.version(), version("0.4.3"));
        assert!(before.version() < middle.version());
        assert_eq!(middle.version(), after.version());
        // The pre-release sorts before the release with the same core version.
        assert!(middle.is_prerelease() && !after.is_prerelease());
    }

    /// Zero padding preserves chronological lexical order.
    #[test]
    fn later_builds_sort_later() {
        let morning = pre(&["v1.0.0"], NOON - 2 * 3600).tag();
        let noon = pre(&["v1.0.0"], NOON).tag();
        assert!(morning < noon, "{morning} should sort below {noon}");
        assert!(morning.contains("09-30-00"), "{morning}");
        assert!(noon.contains("11-30-00"), "{noon}");
    }

    /// The prefix prevents a hash from becoming an all-digit SemVer identifier.
    #[test]
    fn a_commit_is_never_an_all_digit_identifier() {
        let digits = Sha::parse("0012345678901234567890123456789012345678").expect("hex");
        assert_eq!(digits.to_string(), "g0012345");
        assert!(
            !digits.to_string().bytes().all(|b| b.is_ascii_digit()),
            "an all-digit identifier sorts below its alphanumeric siblings"
        );
    }

    /// The base follows release tags only.
    #[test]
    fn the_base_version_is_the_patch_above_the_newest_release() {
        // No release starts from ORIGIN.
        assert_eq!(pre(&[], NOON).version(), version("0.0.1"));
        assert_eq!(pre(&["v0.1.0"], NOON).version(), version("0.1.1"));
        // Tag order does not affect the newest release.
        assert_eq!(pre(&["v0.3.0", "v0.1.0"], NOON).version(), version("0.3.1"));
        // Pre-release tags do not raise the base version.
        assert_eq!(
            pre(&["v0.1.0", "v0.9.9-dev.2026-01-01.00-00-00.gabc1234"], NOON).version(),
            version("0.1.1")
        );
    }

    /// A release uses the version named by its tag; no manifest version is
    /// consulted.
    #[test]
    fn a_release_publishes_the_version_its_tag_names() {
        let released = |tag: &str| {
            resolve(
                Event::Release(Version::parse_tag(tag).expect("a release tag")),
                [],
                stamp(NOON),
                sha(),
            )
        };
        assert_eq!(released("v0.4.2"), Publish::Release(version("0.4.2")));
        assert_eq!(released("v9.9.9"), Publish::Release(version("9.9.9")));
    }

    /// Non-release strings never become `Event::Release`.
    #[test]
    fn only_a_strict_triple_parses_as_a_release() {
        for malformed in [
            "0.4.2",
            "v0.4",
            "v0.4.2.1",
            "v0.04.2",
            "release-0.4.2",
            "",
            // A well-formed pre-release tag is still not a release.
            "v0.4.2-dev.2026-08-25.11-30-00.gabc1234",
        ] {
            assert!(
                Version::parse_tag(malformed).is_none(),
                "{malformed} parsed as a release"
            );
        }
    }

    #[test]
    fn a_tag_round_trips_through_its_parser() {
        for text in ["v0.0.0", "v1.2.3", "v10.20.30"] {
            let parsed = Version::parse_tag(text).expect(text);
            assert_eq!(format!("v{parsed}"), text);
        }
    }

    /// The resolver has no manifest-version input.
    #[test]
    fn the_algebra_takes_no_manifest() {
        let pushed = pre(&["v0.1.0"], NOON);
        assert_eq!(pushed.version(), version("0.1.1"));
    }
}
